### ISSUE TITLE
Dealing with positive and negative infinity as return value for order(Group), index(G,H), exponent(G), degree(poly), ....

### ISSUE THREAD

**Issue by fingolfin** [oscar-system/Oscar.jl#585]
We've had this discussion multiple times in multiple places and situations, but I am not sure we ever recorded a clear outcome, plus I have some new doubts, so, let's try to get this settled (even if this in the end just means someone points me at a different issue/PR/discussion were this was already settled before :joy:).

Many group related functions return either a positive integer or +infinity, which here is meant in the sense of an entity that is comparable with any integer and larger than all of them. Optionally it can also satisfy some rules like `infinity + int = infinity` etc.. (I deliberately do not wish to identify it with any particular cardinality). Here is an incomplete list:
- `order`
- `rank`
- `exponent`
- `index`

There is a similar situation for polynomials: in textbooks, the degree of the zero polynomial is often defined to be `-infinity`. Sometimes it is also deliberately left undefined. Or valuations, where both plus and minus infinity play a role.

After discussions with @ThomasBreuer @wbhart and @thofma this morning, I see the following options:

1. have a variant of the functions (say `isfinite_and_order`) which returns a tuple `(Bool, fmpz)`
2. throw a dedicated exception, say `InfinityException`
3. use a "tagged type", see  https://github.com/kalmarek/GroupsCore.jl/issues/8 and https://gist.github.com/kalmarek/640630a53d5837a5c7fe401eb2f047bd by @kalmarek 
4. use a union with a singleton type as is done in Hecke, see https://github.com/nemocas/Nemo.jl/blob/master/src/Infinity.jl

(Out of completeness, let me mention https://github.com/JuliaMath/Infinities.jl but I don't think we should use it at this point, or perhaps ever, for the usual reasons (lack of matureness, slight mismatch in goals and requirements, general risk of an external dependency).

Here is my personal evaluation of these:
1. IMHO this leads to rather awkward code that is annoying to read 
2. exceptions have a lot of overhead and can make it difficult to follow control flow
3. the resulting objects cannot in general be used in places that expect fmpz/bigint, so usually one of the first things with these will be to unwrap them. Pro: the type prints in a nice way
4. my favorite right now: conceptually easy; meshes well with Julia's union splitting optimization ([see here](https://julialang.org/blog/2018/08/union-splitting/)]

Now, we don't need to pick a single one, different places can use different approaches. But if they use different things, they should have good reasons, IMHO...

Anyway, for the groups stuff, I strongly tend towards approach 4. But if anyone has more thoughts on this or other alternatives, please chime in

**Comment by fingolfin** [comment]
@fieker reminded me about option 5: enhance `fmpz` to allow it to represent infinity (and `-infinity`). See also https://github.com/Nemocas/Nemo.jl/pull/855

**Comment by tthsqe12** [comment]
I have been thinking about this as well. AbstractAlgebra needs an Infinity. Valuation can return +infty, and degree can return -infty. We also need an indeterminate so that we can add, sub, and mul these guys. My initial idea was a `SuperInteger{T}` which holds a normal integer of type `T` and an enum with one of four possible values:
1. the object is positive infinity
2. the object is negative infinity
3. the object is indeterminate
4. the object is the normal integer of type `T`

However, if the union splitting is good idea (exponential explosion in code size?) then we might go with just an infinite object for representing 1, 2 & 3 and have the function return either a normal integer or this inifinite object.

**Comment by fieker** [comment]
On Fri, Nov 26, 2021 at 04:34:28AM -0800, dan wrote:
> I have been thinking about this as well. AbstractAlgebra needs an Infinity. Valuation can return +infty, and degree can return -infty. We also need an indeterminate so that we can add, sub, and mul these guys. My initial idea was a `SuperInteger{T}` which holds a normal integer of type `T` and an enum with one of four possible values:
> 1. the object is positive infinity
> 2. the object is negative infinity
> 3. the object is indeterminate
> 4. the object is the normal integer of type `T`
> 
> However, if the union splitting is good idea (exponential explosion in code size?) then we might go with just an infinite object for representing 1, 2 & 3 and have the function return either a normal integer or this inifinite object.
> 
Cannot this be handled, in Nemo, via mangling of fmpz?

**Comment by tthsqe12** [comment]
libflint expects `Int`s `d` with (1) `|d| < 2^62` or (2) `d > 2^62` with `d =` pointer that has been mangled a certain way. Either we crash libflint with infinite input or check each `fmpz` before passing it to libflint? sounds bad... need another type...

**Comment by thofma** [comment]
Usually we need `Int` + `+/-inf` and not `fmpz` + `+/-inf`. We don't want the degree to return an `fmpz` or the valuation to return an `fmpz`.

I would be in favor of just one infinite object, not parametrized by anything.

**Comment by tthsqe12** [comment]
> I would be in favor of just one infinite object, not parametrized by anything.

So the inferred return of `degree` is Union{Blah, Int}? I cannot see a point in one type just for +infinity. So Blah is one type that represents either +infty, -infty, or nan.

**Comment by thofma** [comment]
If we go with `SuperInteger{T}`, we would need to adjust all signatures to accept also `SuperInteger`? With the `Union` approach, everything would just continue to work.

And yes, one infinity object that can either represent +infty, -infty (why nan?).

**Comment by tthsqe12** [comment]
I agree that SuperInteger{T} is annoying, and I also agree that the Union approach is easy for us (but not for the julia runtime, maybe we will be writing degree(poly)::Int all over the place). However, I am surprised to hear Mr. Type Stable endorsing the type unstable approach. Maybe my FreeAssAlgebra can work the way I wanted it to after all?

**Comment by thofma** [comment]
Mr. Type Stable here. It never crossed my mind to change `degree` to return `-inf` for the zero polynomial. I agree that changing the behaviour could lead to some unfortunate compile time or even run time regressions. So I would rather not change it. For what it is worth, Magma returns `-1` for the degree of the zero polynomial (the same that we are doing right now).

**Comment by fingolfin** [comment]
> And yes, one infinity object that can either represent +infty, -infty

But why??? I can only see this cause pain and suffering in terms of people being completely caught by surprise by this if you call it e.g. `Infinity`. So how would you call it `InfinityOrMinusInfinity` ?

**Comment by tthsqe12** [comment]
Alright, one empty type `Infinity` which prints as `inifinity` in text output, and another empty type `MinusInfinity` which prints as `-infinity`. As a joke, we could also just use Float~64~32 as this supports both infinities :).

**Comment by thofma** [comment]
Since we want to rely on the small union splitting (https://julialang.org/blog/2018/08/union-splitting/), we don't want to return a union of dozens of types. The exact number for which splitting is done is an implementation detail, but we can rely on it being `>= 2` (since iteration in julia is relying on it being `>= 2`).

**Comment by fingolfin** [comment]
Where is the "dozens of types" coming from? If we have `Infinity` and `MinusInfinity` then I am pretty sure in many places it'll just be `Union{Int, Infinity}` or `Union{fmpz, Infinity}`. What are things that need to deal with both plus and minus infinity?

**Comment by thofma** [comment]
Makes sense.

**Comment by fingolfin** [comment]
To make some progress with this, here is some unfinished code implementing something like this... not sure if subtyping `Integer` is brilliant or horrible, though (it sure complicates things in this code snippet!)
```julia
module FooBar

using Nemo: fmpz

abstract type AbstractInfinity <: Integer end
struct Infinity <: AbstractInfinity end
struct MinusInfinity <: AbstractInfinity end

const infinity = Infinity()
export infinity

Base.show(io::IO, ::Infinity) = print(io, "infinity")
Base.show(io::IO, ::MinusInfinity) = print(io, "-infinity")

Base.string(::Infinity) = "infinity"
Base.sign(y::Infinity) = 1
Base.signbit(y::Infinity) = false

Base.string(::MinusInfinity) = "-infinity"
Base.sign(y::MinusInfinity) = -1
Base.signbit(y::MinusInfinity) = true

Base.isone(::AbstractInfinity) = false
Base.iszero(::AbstractInfinity) = false
Base.isinf(::AbstractInfinity) = true
Base.isfinite(::AbstractInfinity) = false

Base.:-(::Infinity) = MinusInfinity()
Base.:-(::MinusInfinity) = Infinity()

Base.:(==)(x::Infinity, y::Infinity) = true
Base.:(==)(x::MinusInfinity, y::MinusInfinity) = true
Base.:(==)(x::Infinity, y::MinusInfinity) = false
Base.:(==)(x::MinusInfinity, y::Infinity) = false

Base.isless(x::AbstractInfinity, y::AbstractInfinity) = sign(x) < sign(y)
Base.:<(x::AbstractInfinity, y::AbstractInfinity) = isless(x, y)
Base.:<=(x::AbstractInfinity, y::AbstractInfinity) = x === y || isless(x, y)

function Base.:+(x::AbstractInfinity, y::AbstractInfinity)
    sign(x) == sign(y) && return x
    throw(ArgumentError("Cannot add infinities with different sign"))
end

function Base.:-(x::AbstractInfinity, y::AbstractInfinity)
    sign(x) != sign(y) && return x
    throw(ArgumentError("Cannot subtract infinities with equal sign"))
end

for T in [Integer, fmpz]
    @eval begin
        Base.:(==)(x::AbstractInfinity, y::$T) = isinf(y) && sign(x) == sign(y)
        Base.:(==)(y::$T, x::AbstractInfinity) = x == y

        Base.isless(x::AbstractInfinity, y::$T) = sign(x) < 0 && y != -infinity
        Base.isless(x::$T, y::AbstractInfinity) = sign(y) > 0 && x != infinity

        Base.:<(x::AbstractInfinity, y::$T) = isless(x, y)
        Base.:<(x::$T, y::AbstractInfinity) = isless(x, y)

        Base.:<=(x::AbstractInfinity, y::$T) = x === y || isless(x, y)
        Base.:<=(x::$T, y::AbstractInfinity) = x === y || isless(x, y)

        Base.:+(::$T, y::AbstractInfinity) = y
        Base.:+(y::AbstractInfinity, ::$T) = y

        Base.:-(y::AbstractInfinity, ::$T) = y
        Base.:-(::$T, y::AbstractInfinity) = -y

        function Base.:*(a::$T, y::AbstractInfinity) 
            iszero(a) && throw(ArgumentError("Cannot multiply $a * $y"))
            a > 0 ? y : (-y)
        end
        Base.:*(y::AbstractInfinity, a::$T) = a * y
    end
end

end


#=


using FooBar, Test

@testset "Infinities" begin
  @test sign(infinity) == 1
  @test sign(-infinity) == -1


  elems = [-infinity, -5, 0, +5, infinity]

  @testset "Infinities: testing operator $(op)" for op in [ !=, ==, <, <=, >, >=, isless ]
    for i in 1:length(elems), j in 1:length(elems)
      x = elems[i]
      y = elems[j]
      @test op(x, y) == op(i, j)
    end
  end

  @test infinity + infinity == infinity
  @test infinity + 5 == infinity
  @test 5 + infinity == infinity

  @test -infinity + -infinity == -infinity
  @test -infinity + 5 == -infinity
  @test 5 + -infinity == -infinity

  @test_throws ArgumentError infinity - infinity
  @test_throws ArgumentError infinity + -infinity
  @test_throws ArgumentError -infinity + infinity


  @test infinity * 5 == infinity
  @test 5 * infinity == infinity
  @test infinity * -5 == -infinity
  @test -5 * infinity == -infinity

  @test -infinity * 5 == -infinity
  @test 5 * -infinity == -infinity

  @test -infinity * -5 == infinity
  @test -5 * -infinity == infinity

  @test  infinity *  infinity ==  infinity
  @test -infinity *  infinity == -infinity
  @test  infinity * -infinity == -infinity
  @test -infinity * -infinity ==  infinity

end


=#
```

**Comment by fingolfin** [comment]
I note that in Hecke there is `PosInf`, which implements parts of this but "just" for plus infinity.

Perhaps we can enhance it by adding `NegInf` (or whatever); also perhaps it then should go to AbstractAlgebra so it can be used more widely?

**Comment by thofma** [comment]
Yes. I am open to renaming everything and your suggestion. I am still a bit unsure about `degree(f::PolyElem)` returning `-inf`. For this I would like to first see some benchmarks (including compilation benchmarks).
